### PR TITLE
Added main thread assertions

### PR DESCRIPTION
**Pull request by tevelee**
Ensuring at runtime that all Backpack UI components are being called from the main thread.

+ [x] Check this if you have read and followed the [contributing guidelines](https://github.com/Skyscanner/backpack-ios/blob/master/CONTRIBUTING.md)

_If you are curious about how we review, please read through the [code review guidelines](./CODE_REVIEW_GUIDELINES.md)_

### CODE REVIEW

File: Backpack/Common.h [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef BPK_Common_h
+#define BPK_Common_h
+
+#import <pthread.h>
+
+#if !defined(NS_BLOCK_ASSERTIONS)
+#define BPK_ASSERTIONS_ENABLED 1
```
Comment (k0nserv) [inline Backpack/Common.h]: Do we need our own method for turning assertions on and off?

File: Backpack/Common.h [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+#else
+#define BPK_ASSERTIONS_ENABLED 0
+#endif
+
+#define BPKAssert(condition, desc, ...) NSAssert(condition, desc, ##__VA_ARGS__)
```
Comment (k0nserv) [inline Backpack/Common.h]: This should also be guarded with a `BPK_ASSERTIONS_ENABLED` check if we do think we should have a custom assertion control?

File: Backpack/Panel/Classes/BPKPanel.m [hunk 3]
```diff
@@ -64,6 +68,7 @@ - (void)setPadded:(BOOL)padded {
 }
 
 - (void)addSubview:(UIView *)view {
+    BPKAssertMainThread();
     NSAssert(self.subviews.count == 0, @"BPKPanel can only have a single subview");
```
Comment (k0nserv) [inline Backpack/Panel/Classes/BPKPanel.m]: Should we switch out `NSAssert`s for `BPKAssert`s?

**Comment by tevelee** [comment]
Thanks @shaundon and @k0nserv for the reviews! I'll update the branch shortly, implementing Hugo's suggestions. 

Regarding Shaun's question, I don't think there's a better way than guarding public methods, manually. 
I took Texture (the old AsyncDisplayKit) as an example, they do it the same way: https://github.com/TextureGroup/Texture/blob/d4efe95ec5399def621096eca9016acdabefaf0b/Source/ASCellNode.mm#L208

**Comment by shaundon** [comment]
Fair enough - thanks for clarifying!

File: Backpack/Common.h [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef BPK_Common_h
+#define BPK_Common_h
+
+#import <pthread.h>
+
+#define BPKAssertMainThread() NSAssert(defined(NS_BLOCK_ASSERTIONS) || 0 != pthread_main_np(), @"This method must be called on the main thread")
```
Comment (k0nserv) [inline Backpack/Common.h]: Do we even need to check `NS_BLOCK_ASSERTIONS` ourselves? My impression is that `NSAssert` already does that

Comment (tevelee) [reply Backpack/Common.h]: Good point

**Pull request merged**